### PR TITLE
Use cargo-deny to audit dependencies

### PR DESCRIPTION
**Pull request by olix0r**
The cargo-deny utility provides a convenient way to audit the proxy's
dependency tree to audit security vulnerabilities & licensing.
Furthermore, it helps detect whether we have multiple versions of a
given dependency.

This change adds a `deny.toml` that configures the project to detect all
such issues and it adds a CI task that can help prevent us from
regressing.

This change introduces a number of dependency updates to eliminate
issues flagged by cargo-deny:

* `futures` has been updated to v0.3.9 to fix safety issues;
* `pin-project` has been updated to v1;
* `regex` has been updated to v1;
* `quickcheck` has been updated to v1, which caught minor issues in
  exp-backoff that have been fixed;

Furthermore, this change updates our github actions to use more more
recent versions of the base container image and updates actions/checkout
to v2.

### CODE REVIEW

File: deny.toml
Comment (hawkw) [inline deny.toml:68]: TIOLI: maybe we should just allow the "tokio", "tower", and "hyperium" orgs globally, to make it less churn  to patch stuff in the future?

File: linkerd/exp-backoff/src/lib.rs
Comment (hawkw) [inline linkerd/exp-backoff/src/lib.rs:70]: i think this is backwards:
```suggestion
            return Err(InvalidBackoff("jitter must be finite"));
```

File: linkerd/stack/src/box_response.rs
Comment (hawkw) [inline linkerd/stack/src/box_response.rs:13]: this was renamed to `BoxFuture` before merging #826, so I think this file is now a duplicate and is unused...can we just remove this?

File: linkerd/stack/src/lib.rs
Comment (hawkw) [inline linkerd/stack/src/lib.rs:28]:
```suggestion
```

File: opencensus-proto/Cargo.toml
Comment (hawkw) [inline opencensus-proto/Cargo.toml:5]: do we know what license the opencensus protobuf API definition is distributed under? does this need to match that?

**Review by hawkw — CHANGES_REQUESTED** [review]
overall, LGTM; I noticed a couple minor issues we should address, and had some additional questions. in particular, let's get rid of `BoxResponse`, which is now unused and i think should have been deleted when merging main, and let's fix the wrong error message in backoff.

File: linkerd/exp-backoff/src/lib.rs
Comment (olix0r) [reply linkerd/exp-backoff/src/lib.rs:70]: no, the jitter must be finite. this error is thrown when it's not finite.

File: opencensus-proto/Cargo.toml
Comment (olix0r) [reply opencensus-proto/Cargo.toml:5]: It matches https://github.com/census-instrumentation/opencensus-proto

File: deny.toml
Comment (olix0r) [reply deny.toml:68]: I'd rather add explicit allowances as we need them. Since we control the linkerd project, though, I'm okay with a blanket allowance for linkerd.

File: linkerd/exp-backoff/src/lib.rs
Comment (olix0r) [reply linkerd/exp-backoff/src/lib.rs:70]: oh i see yes you're right. duh.

**Review by hawkw — APPROVED** [review]
lgtm when CI passes!

**Pull request merged**
